### DISCUSSION TITLE
Setting expire_on_commit=False still causes DetachedInstanceError

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#8731] [Usage Questions] [answered]
```python 
session = sessionmaker(bind=engine, autoflush=False, expire_on_commit=False)

def create_user(self, create: CreateUser) -> User:
    with session.begin() as db:
        create.password = jwt_security.get_hash_password(create.password)
        new_user = User(**create.dict())
        db.add(new_user)
        return new_user
```
After sending the request, it still receives `sqlalchemy.orm.exc.DetachedInstanceError: Instance <User at 0x204e28e91c0> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
`

**Comment by wu-clan** [comment]
The document is here:
- https://docs.sqlalchemy.org/en/20/errors.html#parent-instance-x-is-not-bound-to-a-session-lazy-load-deferred-load-refresh-etc-operation-cannot-proceed
- https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html#closing-a-session

**Comment by zzzeek** [comment, marked as answer]
hi -

the code block you have creates a new Session when you call upon "session.begin()", since the "session" variable is a "sessionmaker()" instance, that is, a factory for new sessions.    When the block exits, the transaction is committed, and the session is closed.   The "new_user" object is returned, now detached as the session would be closed.

If you wanted the create_user() function to return "new_user" while still in an attached state, the caller would have to pass in the Session to which it would like this user object to be attached:

```py

def create_user(self, db: Session, create: CreateUser) -> User:
    create.password = jwt_security.get_hash_password(create.password)
    new_user = User(**create.dict())
    db.add(new_user)
    return new_user

```

The above function is a better way to write functions, as individual functions that work with data need not be concerned with the transactional boundaries, which is a separate concern.  The section at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it addresses this programming issue directly.

**Reply by wu-clan** [reply]
Sorry, ignored the documentation here, great, I'll rewrite the existing design 😭
